### PR TITLE
[dynamo] Be clearer about dict subtype source availability

### PR DESCRIPTION
**Pull request by jon-chuang**
```
# [NOTE] OrderedDict, dict subtypes must always have source
# We cannot instantiate such subtypes in-graph due to builtin __new__
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by ezyang** [comment]
Needs test

**Comment by jon-chuang** [comment]
Added test showing that we cannot create arbitrary dict subtype in graph

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -250,9 +250,12 @@ def call_method(
             if method is object.__init__:
                 return ConstantVariable.create(None)
 
-            if method is collections.OrderedDict.keys and self.source:
+            # [NOTE] OrderedDict, dict subtypes must always have source
+            # We cannot instantiate such subtypes in-graph due to builtin __new__
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:254]: I mean, in principle we could if we implemented their `__new__` correctly, no?

**Review by ezyang — APPROVED** [review]
sure, tighter invariants sound cool to me

File: torch/_dynamo/variables/user_defined.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/user_defined.py:254]: yes, but I didn't want to deal with that case if it is not on usage path.

Comment (jon-chuang) [reply torch/_dynamo/variables/user_defined.py:254]: Instantiating __new__ directly is very rare so we assume it never happens.

**Pull request closed without merging**